File: app/src/main/java/com/example/breezil/favnews/di/module/AppModule.java
package com.example.breezil.favnews.di.module;

import android.app.Application;
import android.arch.persistence.room.Room;

import com.example.breezil.favnews.api.NewsApi;
import com.example.breezil.favnews.db.AppDatabase;
import com.example.breezil.favnews.utils.Constant;
import com.example.breezil.favnews.utils.helpers.LiveDataCallAdapterFactory;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

@Module(includes = ViewModelModule.class)
public class AppModule {
    @Singleton
    @Provides
    NewsApi provideNewsApi() {
        return new Retrofit.Builder()
                .baseUrl(Constant.BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(new LiveDataCallAdapterFactory())
                .build()
                .create(NewsApi.class);
    }
    @Singleton
    @Provides
    AppDatabase provideDb(Application app) {
        return Room.databaseBuilder(app, AppDatabase.class, "favnew.db").build();
    }
}

File: app/src/main/java/com/example/breezil/fave/di/module/AppModule.kt
package com.example.breezil.fave.di.module

import android.app.Application
import android.arch.persistence.room.Room

import com.example.breezil.fave.api.NewsApi
import com.example.breezil.fave.api.OkHttp
import com.example.breezil.fave.db.AppDatabase
import com.example.breezil.fave.utils.Constant

import javax.inject.Singleton

import dagger.Module
import dagger.Provides
import io.reactivex.disposables.CompositeDisposable
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory

@Module(includes = [ViewModelModule::class])
class AppModule {
    @Singleton
    @Provides
    internal fun provideNewsApi(): NewsApi {
        val okHttp = OkHttp()
        return Retrofit.Builder()
                .baseUrl(Constant.BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .client(okHttp.client)
                .build()
                .create(NewsApi::class.java)
    }

    @Singleton
    @Provides
    internal fun provideDb(app: Application): AppDatabase {
        return Room.databaseBuilder(app, AppDatabase::class.java, "favnew.db").build()
    }

    @Provides
    internal fun provideCompositeDisposable(): CompositeDisposable {
        return CompositeDisposable()
    }

}
